### ISSUE TITLE
False positive for shorthand_operator when using simd

### ISSUE THREAD

**Issue by stackotter** [realm/SwiftLint#3603]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

The shorthand_operator rule triggers when normal multiplication is supported between two types yet shorthand multiplication is not.

SIMD doesn't support shorthand multiplication between ```simd_float2``` and ```matrix_float2x2``` (but does support normal multiplication)

For example;

```swift
var uv = simd_float2(0, 0)
let matrix = matrix_float2x2(1)

// Compilation error:
//   Binary operator '*=' cannot be applied to operands of type 'simd_float2' (aka 'SIMD2<Float>') and 'matrix_float2x2' (aka 'simd_float2x2')
uv *= matrix 

// Linting error:
//   Shorthand Operator Violation: Prefer shorthand operators (+=, -=, *=, /=) over doing the operation and assigning. (shorthand_operator)
uv = uv * matrix // this is the only valid way of multiplying uv and matrix together
```

It is very possible for me to just add a disable/enable comment either side. Just thought I should report it because it could get pretty annoying when I have to do this more often.

If someone more knowledgeable than me on how SwiftLint works reckons this would be possible to fix I'm willing to give it a go, but it really doesn't sound like something that SwiftLint could do (from my basic understanding of how it works).

The easiest solution may be to add some kind of configuration setting that allows users to specify types to ignore for the rule (if checking variable types is possible the way SwiftLint works?).

Otherwise it seems like I may just need to either disable/enable it everytime this happens or just disable the rule entirely.

**Comment by stackotter** [comment]
However, I have found possibly a bodge around this.

Putting a space after ```matrix``` does not produce the error. So the line in my example above does not actually trigger the error because of the comment at the end of the line.

For now this is a good enough fix for me, but it probably also counts as a bug (I think it arises from checking if the multiplication is followed by other operations.

**Comment by marcelofabri** [comment]
This is a known limitation of the rule: we don't have enough information about the operands' types, so we can't know whether the shorthand operator is available.

If you're dealing a lot with "non usual" types, I'd suggest either:

1. disabling this rule in the file
2. creating yourself customer shorthand operators. I'm not familiar enough with simd though, so this might not be feasible (you might need to define several overloads for the different matrix dimensions)

**Comment by stackotter** [comment]
Thanks, my solution ended up with just disabling and enabling the rule either side of the affected lines. Because as you correctly guessed it's a bit tricky to add extensions for all of the different possible combinations of dimensions that can be multiplied.
